MAILPOET-5145 Simplify integration test cleanup && MAILPOET-5185

johnolek opened the pull request:
## Description

This PR is an attempt to create more of a clean slate before each integration test is run. This is achieved primarily by deleting all the data in the tables that are part of the plugin before each test is run.

This proved to be much more work than I expected. It turns out many of our tests have assumptions about the IDs of created records, which was the main reason some tests would fail in a full run but would pass when run individually. This made it take a significant amount of time to test some changes because I would have to run the entire suite again. I also ran into some intermittent failures that were very difficult to debug.

### In-memory mount for mysql data
I also added a `tmpfs` mount in 644dff22109ccd452ab7fe9dc79f6f1cffd7a011. It does seem to have sped up test execution times, though not by a huge margin.

Before:
<img width="288" alt="image" src="https://user-images.githubusercontent.com/8656640/230202070-828448fa-bce8-4b41-8650-193f84d2305d.png">

After:
<img width="293" alt="image" src="https://user-images.githubusercontent.com/8656640/230202125-56c7a8d4-5e10-42c4-9d7e-8aec0483315a.png">

For reference, here are some run times from a branch that doesn't include any of the changes in this PR:
<img width="283" alt="image" src="https://user-images.githubusercontent.com/8656640/230202797-1e598a76-0559-470b-a1f7-4f1de039f141.png">

Of course times can vary quite a bit from run to run, so take these with a grain of salt.

### [MAILPOET-5185](https://mailpoet.atlassian.net/browse/MAILPOET-5185) bug discovered and fixed
One of the tests that ended up failing was a legitimate failure and uncovered a bug, [MAILPOET-5185](https://mailpoet.atlassian.net/browse/MAILPOET-5185). I fixed that as part of this PR, otherwise the tests wouldn't pass. This fix ensures that the personal data exporter always includes all custom field data for subscribers (see ticket for more details).

## Code review notes

Although there are many changes, the vast majority are simply removing calls to truncating entities. I didn't remove the helper entirely because a few tests rely on truncating tables within the tests.

In my testing, I found that simply deleting all the rows in all the tables was much faster than truncating the tables. I suspect this may be because we never create all that much data to begin with for any given test. Run locally, the version where I truncated the tables before every test used 320 seconds just for the truncating operations, or about 0.14 seconds per test. I couldn't figure out a way to truncate the tables in a single command.

Simply running a `DELETE FROM table` command instead took a total of around 9 seconds total when executing all 2259 tests. 

## QA notes

The only thing that should require testing is the personal data exporting, MAILPOET-5185.

## Linked PRs

_N/A_

## Linked tickets
https://mailpoet.atlassian.net/browse/MAILPOET-5145
https://mailpoet.atlassian.net/browse/MAILPOET-5185

## After-merge notes

This includes some rather large changes, so I plan on publishing a P2 post summarizing what's changed and what engineers should do differently going forward when it gets merged.

[MAILPOET-5185]: https://mailpoet.atlassian.net/browse/MAILPOET-5185?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

johnolek committed:
Don't stop on first failure

MAILPOET-5145

johnolek committed:
Add database cleaner for integration tests

MAILPOET-5145

johnolek committed:
Don't assume auto increment of 1

MAILPOET-5145

johnolek committed:
Remove unnecessary cleanup code

MAILPOET-5145

JanJakes commented:
@johnolek This is awesome! A tiny test extension and so much cleanup as a result. Really looking forward to having this merged!

> In my testing, I found that simply deleting all the rows in all the tables was much faster than truncating the tables.

Interesting! Maybe it has something to do with resetting auto-increments, indexes, and stuff? Anyway, I think deletes are fine, at least they will force us not to rely on constant IDs in the tests 😊

> I couldn't figure out a way to truncate the tables in a single command.

You could try something like the following, but based on your measurements, I think it will still be slower than deletes:

```sql
SELECT 'TRUNCATE TABLE ' + table_name + ';'
FROM information_schema.tables
WHERE table_name LIKE '<our-table-prefix-here>%'
```

And you could try the same for the deletes to compare with your approach, but I doubt there will be any improvements, as what you are doing now is also just a single roundtrip to the DB.

How much work do you think is left? I see that `integration_test_base` is passing, so you might be close?

johnolek commented:
@JanJakes I do think I'm close, but it's often the last 5% that takes the most amount of time. I'm going to try to power through and finish it today if I can, but some of the tests seem rather tricky.

johnolek committed:
Simplify problematic test

MAILPOET-5145

johnolek committed:
Split up problematic test

MAILPOET-5145

johnolek committed:
Don't assume auto increment of 1

MAILPOET-5145

johnolek committed:
Ensure personal data exporter includes all custom fields

The test testExportSubscriberWithCustomField was only passing before
because we were truncating tables in between tests, so the custom field
id ended up matching the subscriber custom field id coincidentally.

MAILPOET-5145
MAILPOET-5185

johnolek committed:
Extract customer creation/cleanup

This also ensures that cleanup always runs after all integration tests

MAILPOET-5145

johnolek committed:
Remove unnecessary cleanup code

MAILPOET-5145

johnolek committed:
Add missing parent::_after calls

MAILPOET-5145

johnolek committed:
Remove unnecessary truncating

MAILPOET-5145

johnolek committed:
Fix issue with IDs

MAILPOET-5145

johnolek committed:
Add back scheduled task cleanup

MAILPOET-5145

johnolek committed:
Prevent high order IDs from breaking tests

These tests were starting to fail if previous tests had created more
orders than the batch size

MAILPOET-5145

johnolek committed:
Remove apparently unnecessary DB query

Removing this code didn't result in any tests failing, so I can only
assume it isn't required for anything.

We were occasionally seeing primary key constraint violations when
attempting to execute this query.

MAILPOET-5145

johnolek committed:
Clean up after automations too

MAILPOET-5145

johnolek committed:
Remove empty _afters

MAILPOET-5145

johnolek committed:
Add tmpfs mount for faster MySQLing

MAILPOET-5145

johnolek commented:
@JanJakes I believe this should be ready for final review. Thank you so much for all of your help with it! If you think the commits are too messy let me know and I can try to rebase/squash, but with the number of changes I'm worried that might be more of a pain than it's worth.

I would like to explore other low hanging fruit for speeding up these tests. The big thing that stands out to me is the amount of time we spend pulling docker images. It's about a minute for every single job. I tried enabling circleci's docker layer caching in bb5ae12a1f31564d3221f8f249da0a06083ff680, but it didn't seem to work. I'll file a followup ticket for this unless you know of an easy fix, or a reason why we shouldn't do this at all (I know we've had naming conflicts in the past, but I think we might be able to fix that by including more unique data in the container names like in the commit above).

Even though this has taken way longer than I had ever hoped/planned, please don't hesitate to bring up any issues/ideas for improvements, even small ones!

JanJakes commented on mailpoet/lib/DI/ContainerConfigurator.php:
```diff
@@ -436,7 +436,7 @@ public function configure(ContainerBuilder $container) {
     $container->autowire(\MailPoet\Subscription\CaptchaFormRenderer::class)->setPublic(true);
     $container->autowire(\MailPoet\Subscription\Captcha\CaptchaSession::class);
-    $container->autowire(\MailPoet\Subscription\Captcha\CaptchaRenderer::class);
+    $container->autowire(\MailPoet\Subscription\Captcha\CaptchaRenderer::class)->setPublic(true);
```
This looks a bit random, does it belong here?

JanJakes commented on mailpoet/tests/_support/IntegrationCleanupExtension.php:
```diff
@@ -0,0 +1,67 @@
+      'mailpoet_automation_triggers',
+      'mailpoet_automation_versions',
+      'mailpoet_automations',
+    ];
```
When we add a new automation table, we'll certainly forget to add it here. Could we match everything starting with `mailpoet_automation`, maybe? But if we do that... then why not matching all that starts with `mailpoet_` (excluding `mailpoet_migrations`).

JanJakes replied:
Done in: https://github.com/mailpoet/mailpoet/pull/4818/commits/03396e48d5e8cc4c16c408ae9852e85c6ddda86f

JanJakes commented on mailpoet/tests/integration/AutomaticEmails/WooCommerce/Events/PurchasedProductTest.php:
```diff
@@ -284,14 +282,7 @@ private function createWooSubscriber(): SubscriberEntity {
-    $this->truncateEntity(ScheduledTaskEntity::class);
-    $this->truncateEntity(ScheduledTaskSubscriberEntity::class);
+    parent::_after();
     WPFunctions::set(new WPFunctions);
```
Maybe we could add this logic to the bootstrap or cleanup extension? I see it repeated and the runtime cost of this will probably be close to nothing...

JanJakes replied:
Done in: https://github.com/mailpoet/mailpoet/pull/4818/commits/0ef10c0348e845a09939afe795299897ae790897

JanJakes commented on mailpoet/tests/integration/Automation/Integrations/MailPoet/Triggers/SomeoneSubscribesTriggerTest.php:
```diff
@@ -71,6 +71,7 @@ public function dataForTestTriggeredByAutomationRun(): array {
   public function _after() {
+    parent::_after();
     $segmentIds = $this->getSegmentIds(array_keys($this->segments));
     $this->segmentRepository->bulkDelete($segmentIds);
```
Is this needed now, when we delete everything?

JanJakes replied:
Done in: https://github.com/mailpoet/mailpoet/pull/4818/commits/557577a432821dcc8238308bb4ff7b349b3907a4

JanJakes commented on mailpoet/tests/integration/Cron/ActionScheduler/Actions/DaemonRunTest.php:
```diff
@@ -127,7 +126,5 @@ private function cleanup(): void {
     $wpdb->query('TRUNCATE ' . $actionsTable);
     $claimsTable = $wpdb->prefix . 'actionscheduler_claims';
     $wpdb->query('TRUNCATE ' . $claimsTable);
```
I wonder if we could try to include Action Scheduler tables in the cleanup extension 🤔

JanJakes commented on mailpoet/tests/integration/Cron/Workers/SchedulerTest.php:
```diff
@@ -981,17 +974,7 @@ private function getScheduler(?SubscribersFinder $subscribersFinder = null): Sch
 
   public function _after() {
+    parent::_after();
     Carbon::setTestNow();
```
Another candidate to be also extracted to bootstrap/cleanup extension?

JanJakes replied:
Done in: https://github.com/mailpoet/mailpoet/pull/4818/commits/5bffd812189393f85ebf2985aaf32c8fe1134286

JanJakes commented on mailpoet/tests/integration/Models/CustomFieldTest.php:
```diff
@@ -145,6 +145,7 @@ public function testItCanHaveAValue() {
+    parent::_after();
     CustomField::deleteMany();
     Subscriber::deleteMany();
     SubscriberCustomField::deleteMany();
```
I think these deletes are also not needed anymore.

JanJakes replied:
Done in: https://github.com/mailpoet/mailpoet/pull/4818/commits/557577a432821dcc8238308bb4ff7b349b3907a4

JanJakes commented on mailpoet/tests/integration/Models/SubscriberCustomFieldTest.php:
```diff
@@ -77,6 +77,7 @@ public function testItCanDeleteSubscriberRelations() {
 
   public function _after() {
+    parent::_after();
     SubscriberCustomField::deleteMany();
```
This one doesn't seem to be needed anymore.

JanJakes replied:
Done in: https://github.com/mailpoet/mailpoet/pull/4818/commits/557577a432821dcc8238308bb4ff7b349b3907a4

JanJakes commented on mailpoet/tests/integration/Models/SubscriberSegmentTest.php:
```diff
@@ -336,6 +336,7 @@ public function testItCannotDeleteSubscriptionToWPAndWooCommerceSegments() {
+    parent::_after();
     Segment::deleteMany();
     Subscriber::deleteMany();
     SubscriberSegment::deleteMany();
```
And these, I think, are not needed anymore.

JanJakes replied:
Done in: https://github.com/mailpoet/mailpoet/pull/4818/commits/557577a432821dcc8238308bb4ff7b349b3907a4

JanJakes commented on mailpoet/tests/integration/Newsletter/ShortcodesTest.php:
```diff
@@ -482,15 +482,11 @@ public function _createQueue(): SendingQueueEntity {
-    $this->truncateEntity(CustomFieldEntity::class);
-    $this->truncateEntity(SubscriberCustomFieldEntity::class);
     if ($this->wPPost) wp_delete_post($this->wPPost, true);
     if ($this->wPUser) wp_delete_user($this->wPUser->ID);
```
What if we also delete all posts and related metadata in the extension? As long as we append the deletes to the same long multi-query, I think it would be just a tiny cost...?

JanJakes commented on mailpoet/tests/integration/Newsletter/ViewInBrowser/ViewInBrowserControllerTest.php:
```diff
@@ -237,10 +237,7 @@ public function testItResetsQueueForAutomaticEmailsInPreview() {
+    parent::_after();
     // reset WP user role
     $wpUser = wp_get_current_user();
     $wpUser->add_role('administrator');
```
And when at it, maybe even resetting the WP users table to some defaults can be useful. I'm sharing these thoughts as I see them, but I think we can extract all of that into a follow-up ticket.

JanJakes commented on mailpoet/tests/integration/Segments/SegmentSubscribersRepositoryTest.php:
```diff
@@ -247,10 +247,6 @@ private function createDynamicSegmentEntity(string $role = 'editor'): SegmentEnt
-    $this->truncateEntity(SubscriberEntity::class);
-    $this->truncateEntity(SubscriberSegmentEntity::class);
-    $this->truncateEntity(DynamicSegmentFilterEntity::class);
     $this->clearSubscribersCountCache();
```
Where is the cache stored, is it a WP transient? Just thinking if we should add transients to our list of additional things to clean up. On the other hand, they may also stay in memory 🤔

JanJakes replied:
Ahh, we call that already in `_bootstrap` so removed: https://github.com/mailpoet/mailpoet/pull/4818/commits/557577a432821dcc8238308bb4ff7b349b3907a4

JanJakes commented on mailpoet/tests/integration/Subscribers/ImportExport/Import/ImportTest.php:
```diff
@@ -880,12 +879,7 @@ private function createSubscriber(
-    $this->truncateEntity(SegmentEntity::class);
-    $this->truncateEntity(SubscriberSegmentEntity::class);
+    parent::_after();
     $this->truncateEntity(CustomFieldEntity::class);
```
Did the custom field entity truncation stay here by mistake?

JanJakes replied:
Removed (https://github.com/mailpoet/mailpoet/pull/4818/commits/557577a432821dcc8238308bb4ff7b349b3907a4) and fixed a related bug: https://github.com/mailpoet/mailpoet/pull/4818/commits/550ac60c5e650fe496d1a8084987df28d81e9d65

JanJakes reviewed (commented):
Great job, this is really awesome! 👏 

> If you think the commits are too messy let me know and I can try to rebase/squash, but with the number of changes I'm worried that might be more of a pain than it's worth.

Yeah, no need to do something overly complex. There are some try-revert sequences, so we could maybe drop those, if reasonably easy.

> I would like to explore other low hanging fruit for speeding up these tests. The big thing that stands out to me is the amount of time we spend pulling docker images. It's about a minute for every single job. I tried enabling circleci's docker layer caching in https://github.com/mailpoet/mailpoet/commit/bb5ae12a1f31564d3221f8f249da0a06083ff680, but it didn't seem to work. I'll file a followup ticket for this unless you know of an easy fix, or a reason why we shouldn't do this at all (I know we've had naming conflicts in the past, but I think we might be able to fix that by including more unique data in the container names like in the commit above).

As for Docker layer caching, [there was a thread](https://a8c.slack.com/archives/C01H71ZLBGQ/p1667985836388249) pretty much deciding against it, as it's an expensive feature that brings less speed-up than one may think. Docker layer caching makes sense only when `docker build` or `docker compose build` is actually run on the CI, which I think, is something we don't do. For prebuilt images, I wouldn't expect any reasonable speed ups (the cache is AWS S3, the Docker hub image probably as well).

We can certainly try to optimize this in some ways, so a ticket may be useful in any case.

> Even though this has taken way longer than I had ever hoped/planned, please don't hesitate to bring up any issues/ideas for improvements, even small ones!

I'm dropping all the ideas that pop up when looking at the PR. Of course, we can extract most of them (if not all) to follow-up tickets. If my support load is not high, I may try to look at some of those today. If I touch anything, I'll let you know here and/or on Slack.

JanJakes commented on mailpoet/tests/integration/_bootstrap.php:
```diff
@@ -98,8 +98,9 @@ public function setUp(): void {
     $this->entityManager = $this->diContainer->get(EntityManager::class);
-    // Cleanup scheduled tasks from previous tests and switch cron to Linux method
+    // Cleanup scheduled tasks from previous tests
     $this->truncateEntity(ScheduledTaskEntity::class);
```
This was removed and then added back. Is there a reason to keep this? I think it should be safe to remove.

JanJakes replied:
Removed: https://github.com/mailpoet/mailpoet/pull/4818/commits/557577a432821dcc8238308bb4ff7b349b3907a4

JanJakes committed:
Clean all mailpoet tables by prefix

MAILPOET-5145

JanJakes committed:
Remove more no longer necessary test cleanups

MAILPOET-5145

JanJakes committed:
Extract WPFunctions reset to test bootstrap

MAILPOET-5145

JanJakes committed:
Replace hardcoded custom field ID with the actual one

MAILPOET-5145

JanJakes committed:
Extract Carbon datetime reset to test bootstrap

MAILPOET-5145

JanJakes commented:
@johnolek I had some time this morning, so I did the following:

1. I dropped all the try-revert commits and did a fixup of some like fix of Prettier formatting. I repushed the branch without any other changes. All my additions are added as new commits.
2. Cleaning up all MP tables (except for migrations) by prefix to avoid forgetting this for new automation tables.
3. Removed some forgotten test cleanup, and fixed a test bug that the removal of one more `TRUNCATE` revealed.
4. Extracted `WPFunctions` and `Carbon` reset to `_bootstrap.php`.

What I suggested but _didn't do_ is:
1. Clean up also WP tables (and related in-memory data like transients, if needed).
2. Clean up also Woo tables.
3. Clean up also Action Scheduler tables.

I think we can extract these to a new ticket. We need to be careful not to introduce slowdowns when adding more cleanups. The current speedup looks just great!

I'm hitting approve from my side, but assigning it back to you, @johnolek, so you can verify my changes. Feel free to pass this on to QA, if it looks good to you.

johnolek committed:
More accurate name for cleanup property

MAILPOET-5145

johnolek commented:
@JanJakes Thank you so much for your help with this! I think your changes all look great, and I definitely appreciate you taking care of some of the git cleanup.

I pushed one minor naming change (b3fc97fec814bbe423d59840caca18e1c97ef23e) now that the SQL we're executing isn't just delete statements, otherwise I think this looks good to go! I'll write a followup ticket for the other items you mentioned.

Aschepikov reviewed (approved):
GJ! @johnolek I made a regression for export personal data. I see there are a lot of changes in the pr. Can I merge it, or do you need to do something before it?

johnolek commented:
@Aschepikov Thanks for the review! I would like to publish a P2 before merging, so I can handle the merge.
